File: uzum/php/tests/test_uzum.php
<?php
/**
 * uzumHandle* funksiyalari uchun testlar.
 *
 * Ishga tushirish (loyiha ildizidan):
 *
 *     php tests/test_uzum.php
 *
 * PHPUnit kerak emas — oddiy skript.
 */

// --- Hodisalarni kuzatish ----------------------------------------------------
// MUHIM: bularni uzum_orders.php dan OLDIN e'lon qilamiz. U yerdagi
// funksiyalar `if (!function_exists(...))` bilan o'ralgani uchun bizniki
// kuchda qoladi.

$GLOBALS['confirmed_calls'] = array();
$GLOBALS['reversed_calls'] = array();

function uzumOnConfirmed(UzumTransaction $tx)
{
    $GLOBALS['confirmed_calls'][] = $tx->transId;
}

function uzumOnReversed(UzumTransaction $tx)
{
    $GLOBALS['reversed_calls'][] = $tx->transId;
}

putenv('UZUM_DB_PATH=:memory:');
require_once __DIR__ . '/../uzum_methods.php';

const SERVICE_ID = 101202;
const LOGIN = 'myLogin';
const SECRET = 'myPassword';

uzumSetConfig(array('service_id' => SERVICE_ID, 'webhook_login' => LOGIN, 'webhook_secret' => SECRET));

function tAuth()
{
    return 'Basic ' . base64_encode(LOGIN . ':' . SECRET);
}

function tBadAuth()
{
    return 'Basic ' . base64_encode('wrong:wrong');
}

function tCheckReq($account = '42', array $over = array())
{
    $data = array('serviceId' => SERVICE_ID, 'timestamp' => 1, 'params' => array('account' => $account));
    return array_merge($data, $over);
}

function tCreateReq($transId, $account = '42', $amount = 2500000, array $over = array())
{
    $data = array(
        'serviceId' => SERVICE_ID, 'timestamp' => 1, 'transId' => $transId,
        'params' => array('account' => $account), 'amount' => $amount,
    );
    return array_merge($data, $over);
}

function tConfirmReq($transId, array $over = array())
{
    $data = array(
        'serviceId' => SERVICE_ID, 'timestamp' => 1, 'transId' => $transId,
        'paymentSource' => 'UZCARD', 'phone' => '998901234567',
    );
    return array_merge($data, $over);
}

function tReverseReq($transId, array $over = array())
{
    $data = array('serviceId' => SERVICE_ID, 'timestamp' => 1, 'transId' => $transId);
    return array_merge($data, $over);
}

// --- Kichik test runner ------------------------------------------------------

$GLOBALS['tests_run'] = 0;
$GLOBALS['tests_failed'] = 0;

function check($condition, $name, $info = '')
{
    $GLOBALS['tests_run']++;
    if ($condition) {
        echo "  OK   {$name}\n";
    } else {
        $GLOBALS['tests_failed']++;
        echo "  XATO {$name}" . ($info !== '' ? " -> {$info}" : '') . "\n";
    }
}

function checkEq($expected, $actual, $name)
{
    check($expected === $actual, $name, 'kutilgan: ' . var_export($expected, true) . ', kelgan: ' . var_export($actual, true));
}

// =============================================================================
//  AUTH
// =============================================================================

echo "\nAUTH\n";
uzumDemoCreateOrder('42', 2500000);

list($status, $body) = uzumHandleCheck(tCheckReq(), tBadAuth());
checkEq(400, $status, 'soxta auth -> HTTP 400');
checkEq('10001', $body['errorCode'], 'soxta auth -> 10001');

list($status, $body) = uzumHandleCheck(tCheckReq(), null);
checkEq(400, $status, "auth yo'q -> HTTP 400");
checkEq('10001', $body['errorCode'], "auth yo'q -> 10001");

// =============================================================================
//  /check
// =============================================================================

echo "\n/check\n";

list($status, $body) = uzumHandleCheck(tCheckReq(), tAuth());
checkEq(200, $status, 'check muvaffaqiyatli');
checkEq('OK', $body['status'], 'check status OK');

list($status, $body) = uzumHandleCheck(tCheckReq('999'), tAuth());
checkEq(400, $status, 'check topilmagan');
checkEq('10007', $body['errorCode'], 'check topilmagan -> 10007');

list($status, $body) = uzumHandleCheck(tCheckReq('42', array('serviceId' => 555)), tAuth());
checkEq('10006', $body['errorCode'], "check noto'g'ri serviceId -> 10006");

$req = tCheckReq();
unset($req['timestamp']);
list($status, $body) = uzumHandleCheck($req, tAuth());
checkEq('10005', $body['errorCode'], "check timestamp yo'q -> 10005");

// =============================================================================
//  /create
// =============================================================================

echo "\n/create\n";

list($status, $body) = uzumHandleCreate(tCreateReq('t1'), tAuth());
checkEq(200, $status, 'create muvaffaqiyatli');
checkEq('CREATED', $body['status'], 'create status CREATED');
checkEq('t1', $body['transId'], "create transId to'g'ri");

list($status, $body) = uzumHandleCreate(tCreateReq('t1'), tAuth());
checkEq(400, $status, 'takroriy create -> HTTP 400');
checkEq('10010', $body['errorCode'], 'takroriy create -> 10010 (idempotent EMAS)');

list($status, $body) = uzumHandleCreate(tCreateReq('t2', '42', 100), tAuth());
checkEq('10011', $body['errorCode'], "create noto'g'ri summa -> 10011");

list($status, $body) = uzumHandleCreate(tCreateReq('t3', '999'), tAuth());
checkEq('10007', $body['errorCode'], "create topilmagan hisob -> 10007");

// =============================================================================
//  /confirm
// =============================================================================

echo "\n/confirm\n";

uzumDemoCreateOrder('50', 1000000, 'Daftar');
uzumHandleCreate(tCreateReq('c1', '50', 1000000), tAuth());

list($status, $body) = uzumHandleConfirm(tConfirmReq('c1'), tAuth());
checkEq(200, $status, 'confirm muvaffaqiyatli');
checkEq('CONFIRMED', $body['status'], 'confirm status CONFIRMED');
check(in_array('c1', $GLOBALS['confirmed_calls']), 'confirm uzumOnConfirmed() ni chaqiradi');

list($status, $body) = uzumHandleConfirm(tConfirmReq('c1'), tAuth());
checkEq(400, $status, 'takroriy confirm -> HTTP 400');
checkEq('10016', $body['errorCode'], 'takroriy confirm -> 10016 (idempotent EMAS)');
checkEq(1, count(array_keys($GLOBALS['confirmed_calls'], 'c1')), 'takroriy confirm uzumOnConfirmed ni QAYTA chaqirmaydi');

list($status, $body) = uzumHandleConfirm(tConfirmReq('yoq-id'), tAuth());
checkEq('10014', $body['errorCode'], 'confirm topilmagan -> 10014');

$req = tConfirmReq('c1');
unset($req['paymentSource']);
list($status, $body) = uzumHandleConfirm($req, tAuth());
checkEq('10005', $body['errorCode'], "confirm paymentSource yo'q -> 10005");

// =============================================================================
//  /reverse
// =============================================================================

echo "\n/reverse\n";

uzumDemoCreateOrder('60', 500000, 'Ruchka');
uzumHandleCreate(tCreateReq('r1', '60', 500000), tAuth());

list($status, $body) = uzumHandleReverse(tReverseReq('r1'), tAuth());
checkEq(200, $status, 'CREATED holatidan reverse');
checkEq('REVERSED', $body['status'], 'reverse status REVERSED');
check(in_array('r1', $GLOBALS['reversed_calls']), 'reverse uzumOnReversed() ni chaqiradi');

list($status, $body) = uzumHandleConfirm(tConfirmReq('r1'), tAuth());
checkEq('10015', $body['errorCode'], 'bekor qilingandan keyin confirm -> 10015');

list($status, $body) = uzumHandleReverse(tReverseReq('r1'), tAuth());
checkEq('10018', $body['errorCode'], 'takroriy reverse -> 10018');

// refund (CONFIRMED holatidan reverse)
uzumDemoCreateOrder('70', 700000, 'Qalam');
uzumHandleCreate(tCreateReq('r2', '70', 700000), tAuth());
uzumHandleConfirm(tConfirmReq('r2'), tAuth());

list($status, $body) = uzumHandleReverse(tReverseReq('r2'), tAuth());
checkEq(200, $status, 'CONFIRMED holatidan reverse (refund)');
checkEq('REVERSED', $body['status'], 'refund status REVERSED');
check(in_array('r2', $GLOBALS['reversed_calls']), 'refund uzumOnReversed() ni chaqiradi');

list($status, $body) = uzumHandleReverse(tReverseReq('yoq-id'), tAuth());
checkEq('10014', $body['errorCode'], 'reverse topilmagan -> 10014');

// =============================================================================
//  /status
// =============================================================================

echo "\n/status\n";

uzumDemoCreateOrder('80', 200000, 'Marker');
uzumHandleCreate(tCreateReq('s1', '80', 200000), tAuth());

list($status, $body) = uzumHandleStatus(tReverseReq('s1'), tAuth());
checkEq(200, $status, 'status CREATED');
checkEq('CREATED', $body['status'], 'status CREATED qiymati');
check(!isset($body['confirmTime']), 'status CREATED holatida confirmTime yo\'q');

uzumHandleConfirm(tConfirmReq('s1'), tAuth());
list($status, $body) = uzumHandleStatus(tReverseReq('s1'), tAuth());
checkEq('CONFIRMED', $body['status'], 'status CONFIRMED qiymati');
check(isset($body['confirmTime']), 'status CONFIRMED holatida confirmTime bor');

list($status, $body) = uzumHandleStatus(tReverseReq('yoq-id'), tAuth());
checkEq('10014', $body['errorCode'], 'status topilmagan -> 10014');

// =============================================================================
//  Poyga himoyasi — eng muhim shart
// =============================================================================

echo "\nPOYGA HIMOYASI\n";

uzumDemoCreateOrder('90', 300000, 'Flomaster');
uzumHandleCreate(tCreateReq('p1', '90', 300000), tAuth());

check(uzumMarkConfirmed('p1') !== null, 'uzumMarkConfirmed birinchi marta yozuv qaytaradi');
check(uzumMarkConfirmed('p1') === null, "uzumMarkConfirmed ikkinchi marta NULL qaytaradi (mahsulot ikki marta berilmaydi)");

// =============================================================================

echo "\n";
echo "=====================================\n";
printf("  Jami: %d, xato: %d\n", $GLOBALS['tests_run'], $GLOBALS['tests_failed']);
echo "=====================================\n";

exit($GLOBALS['tests_failed'] > 0 ? 1 : 0);

File: uzum/python/tests/test_uzum.py
"""Uzum Bank webhook'lari uchun testlar.

Ishga tushirish (loyiha ildizidan):

    pip install pytest
    python -m pytest -v
"""

from __future__ import annotations

import base64
import os

import pytest

os.environ.setdefault("UZUM_DB_PATH", ":memory:")

from uzum_payment import handle_check, handle_confirm, handle_create, handle_reverse, handle_status
from uzum_payment.uzum_config import UzumConfig, set_config
from uzum_payment.uzum_orders import demo_create_order, reset_db_for_tests

SERVICE_ID = 101202
LOGIN = "myLogin"
SECRET = "myPassword"

AUTH = "Basic " + base64.b64encode(f"{LOGIN}:{SECRET}".encode()).decode()
BAD_AUTH = "Basic " + base64.b64encode(b"wrong:wrong").decode()


@pytest.fixture(autouse=True)
def setup():
    set_config(UzumConfig(service_id=SERVICE_ID, webhook_login=LOGIN, webhook_secret=SECRET))
    os.environ["UZUM_DB_PATH"] = ":memory:"
    reset_db_for_tests()
    yield
    reset_db_for_tests()


def check_req(account="42", **over):
    data = {"serviceId": SERVICE_ID, "timestamp": 1, "params": {"account": account}}
    data.update(over)
    return data


def create_req(trans_id, account="42", amount=2500000, **over):
    data = {
        "serviceId": SERVICE_ID,
        "timestamp": 1,
        "transId": trans_id,
        "params": {"account": account},
        "amount": amount,
    }
    data.update(over)
    return data


def confirm_req(trans_id, **over):
    data = {
        "serviceId": SERVICE_ID,
        "timestamp": 1,
        "transId": trans_id,
        "paymentSource": "UZCARD",
        "phone": "998901234567",
    }
    data.update(over)
    return data


def reverse_req(trans_id, **over):
    data = {"serviceId": SERVICE_ID, "timestamp": 1, "transId": trans_id}
    data.update(over)
    return data


# --- AUTH ---------------------------------------------------------------


def test_bad_auth_rejected():
    demo_create_order("42", 2500000)
    status, body = handle_check(check_req(), BAD_AUTH)
    assert status == 400
    assert body["errorCode"] == "10001"


def test_missing_auth_rejected():
    demo_create_order("42", 2500000)
    status, body = handle_check(check_req(), None)
    assert status == 400
    assert body["errorCode"] == "10001"


# --- /check ---------------------------------------------------------------


def test_check_success():
    demo_create_order("42", 2500000)
    status, body = handle_check(check_req(), AUTH)
    assert status == 200
    assert body["status"] == "OK"
    assert body["serviceId"] == SERVICE_ID


def test_check_account_not_found():
    status, body = handle_check(check_req(account="999"), AUTH)
    assert status == 400
    assert body["errorCode"] == "10007"


def test_check_wrong_service_id():
    demo_create_order("42", 2500000)
    status, body = handle_check(check_req(serviceId=555), AUTH)
    assert status == 400
    assert body["errorCode"] == "10006"


def test_check_missing_timestamp():
    demo_create_order("42", 2500000)
    data = check_req()
    del data["timestamp"]
    status, body = handle_check(data, AUTH)
    assert status == 400
    assert body["errorCode"] == "10005"


def test_check_already_paid():
    demo_create_order("42", 2500000)
    trans_id = "t1"
    handle_create(create_req(trans_id), AUTH)
    handle_confirm(confirm_req(trans_id), AUTH)

    status, body = handle_check(check_req(), AUTH)
    assert status == 400
    assert body["errorCode"] == "10008"


# --- /create ----------------------------------------------------------------


def test_create_success():
    demo_create_order("42", 2500000)
    status, body = handle_create(create_req("t1"), AUTH)
    assert status == 200
    assert body["status"] == "CREATED"
    assert body["transId"] == "t1"
    assert "transTime" in body


def test_create_duplicate_returns_error_not_idempotent():
    """Uzum Bank hujjati: takroriy /create XATO qaytaradi, muvaffaqiyat emas."""
    demo_create_order("42", 2500000)
    handle_create(create_req("t1"), AUTH)

    status, body = handle_create(create_req("t1"), AUTH)
    assert status == 400
    assert body["errorCode"] == "10010"


def test_create_wrong_amount():
    demo_create_order("42", 2500000)
    status, body = handle_create(create_req("t1", amount=100), AUTH)
    assert status == 400
    assert body["errorCode"] == "10011"


def test_create_account_not_found():
    status, body = handle_create(create_req("t1", account="999"), AUTH)
    assert status == 400
    assert body["errorCode"] == "10007"


def test_create_second_transaction_for_same_order_rejected():
    """Bitta buyurtmaga ikkita PARALLEL faol tranzaksiya ochilmasin."""
    demo_create_order("42", 2500000)
    handle_create(create_req("t1"), AUTH)

    status, body = handle_create(create_req("t2"), AUTH)
    assert status == 400
    assert body["errorCode"] == "10008"


# --- /confirm ---------------------------------------------------------------


def test_confirm_success_calls_on_confirmed(monkeypatch):
    calls = []
    import uzum_payment.uzum_orders as orders_mod

    monkeypatch.setattr(orders_mod, "on_confirmed", lambda tx: calls.append(tx.trans_id))

    demo_create_order("42", 2500000)
    handle_create(create_req("t1"), AUTH)

    status, body = handle_confirm(confirm_req("t1"), AUTH)
    assert status == 200
    assert body["status"] == "CONFIRMED"
    assert calls == ["t1"]


def test_confirm_duplicate_returns_error_not_idempotent():
    """Uzum Bank hujjati: takroriy /confirm XATO qaytaradi."""
    demo_create_order("42", 2500000)
    handle_create(create_req("t1"), AUTH)
    handle_confirm(confirm_req("t1"), AUTH)

    status, body = handle_confirm(confirm_req("t1"), AUTH)
    assert status == 400
    assert body["errorCode"] == "10016"


def test_confirm_unknown_transaction():
    status, body = handle_confirm(confirm_req("yoq"), AUTH)
    assert status == 400
    assert body["errorCode"] == "10014"


def test_confirm_missing_payment_source():
    demo_create_order("42", 2500000)
    handle_create(create_req("t1"), AUTH)
    data = confirm_req("t1")
    del data["paymentSource"]
    status, body = handle_confirm(data, AUTH)
    assert status == 400
    assert body["errorCode"] == "10005"


def test_confirm_after_reverse_fails():
    demo_create_order("42", 2500000)
    handle_create(create_req("t1"), AUTH)
    handle_reverse(reverse_req("t1"), AUTH)

    status, body = handle_confirm(confirm_req("t1"), AUTH)
    assert status == 400
    assert body["errorCode"] == "10015"


def test_confirm_timeout_auto_cancels(monkeypatch):
    """30 daqiqadan keyin /confirm kelsa — avtomatik bekor qilinadi."""
    demo_create_order("42", 2500000)
    handle_create(create_req("t1"), AUTH)

    import uzum_payment.uzum_methods as methods_mod
    import uzum_payment.uzum_orders as orders_mod

    real_now = methods_mod._now_ms()
    monkeypatch.setattr(
        methods_mod, "_now_ms", lambda: real_now + orders_mod.TRANSACTION_TIMEOUT_MS + 1000
    )

    status, body = handle_confirm(confirm_req("t1"), AUTH)
    assert status == 400
    assert body["errorCode"] == "10015"

    status, body = handle_status(reverse_req("t1"), AUTH)
    assert body["status"] == "REVERSED"


# --- /reverse -----------------------------------------------------------


def test_reverse_from_created():
    demo_create_order("42", 2500000)
    handle_create(create_req("t1"), AUTH)

    status, body = handle_reverse(reverse_req("t1"), AUTH)
    assert status == 200
    assert body["status"] == "REVERSED"


def test_reverse_from_confirmed_is_refund(monkeypatch):
    calls = []
    import uzum_payment.uzum_orders as orders_mod

    monkeypatch.setattr(orders_mod, "on_reversed", lambda tx: calls.append(tx.trans_id))

    demo_create_order("42", 2500000)
    handle_create(create_req("t1"), AUTH)
    handle_confirm(confirm_req("t1"), AUTH)

    status, body = handle_reverse(reverse_req("t1"), AUTH)
    assert status == 200
    assert body["status"] == "REVERSED"
    assert calls == ["t1"]


def test_reverse_duplicate_returns_error():
    demo_create_order("42", 2500000)
    handle_create(create_req("t1"), AUTH)
    handle_reverse(reverse_req("t1"), AUTH)

    status, body = handle_reverse(reverse_req("t1"), AUTH)
    assert status == 400
    assert body["errorCode"] == "10018"


def test_reverse_unknown_transaction():
    status, body = handle_reverse(reverse_req("yoq"), AUTH)
    assert status == 400
    assert body["errorCode"] == "10014"


def test_reverse_respects_can_reverse_hook(monkeypatch):
    import uzum_payment.uzum_orders as orders_mod

    monkeypatch.setattr(orders_mod, "can_reverse", lambda tx: False)

    demo_create_order("42", 2500000)
    handle_create(create_req("t1"), AUTH)
    handle_confirm(confirm_req("t1"), AUTH)

    status, body = handle_reverse(reverse_req("t1"), AUTH)
    assert status == 400
    assert body["errorCode"] == "10017"


# --- /status ------------------------------------------------------------


def test_status_created():
    demo_create_order("42", 2500000)
    handle_create(create_req("t1"), AUTH)

    status, body = handle_status(reverse_req("t1"), AUTH)
    assert status == 200
    assert body["status"] == "CREATED"
    assert "confirmTime" not in body
    assert "reverseTime" not in body


def test_status_confirmed():
    demo_create_order("42", 2500000)
    handle_create(create_req("t1"), AUTH)
    handle_confirm(confirm_req("t1"), AUTH)

    status, body = handle_status(reverse_req("t1"), AUTH)
    assert status == 200
    assert body["status"] == "CONFIRMED"
    assert "confirmTime" in body


def test_status_unknown_transaction():
    status, body = handle_status(reverse_req("yoq"), AUTH)
    assert status == 400
    assert body["errorCode"] == "10014"


# --- Callback xatosi ----------------------------------------------------


def test_on_confirmed_error_does_not_break_response(monkeypatch):
    """on_confirmed xato bersa ham Uzum'ga CONFIRMED ketadi (pul allaqachon yechilgan)."""
    import uzum_payment.uzum_orders as orders_mod

    def broken(tx):
        raise RuntimeError("baza yiqildi")

    monkeypatch.setattr(orders_mod, "on_confirmed", broken)

    demo_create_order("42", 2500000)
    handle_create(create_req("t1"), AUTH)

    status, body = handle_confirm(confirm_req("t1"), AUTH)
    assert status == 200
    assert body["status"] == "CONFIRMED"
